pkg/controller/useraccount/useraccount_controller.go

```diff
@@ -98,75 +99,74 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
-		_ = r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, err.Error())
+	if created, err = r.ensureMapping(reqLogger, userAcc, user, identity); err != nil || created {
 		return reconcile.Result{}, err
-	} else if created {
-		return reconcile.Result{}, r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, "")
 	}
 
 	return reconcile.Result{}, r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioned, "")
 }
 
-func (r *ReconcileUserAccount) ensureUser(userAcc *toolchainv1alpha1.UserAccount) (*userv1.User, bool, error) {
+func (r *ReconcileUserAccount) ensureUser(logger logr.Logger, userAcc *toolchainv1alpha1.UserAccount) (*userv1.User, bool, error) {
 	user := &userv1.User{}
 	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: userAcc.Name}, user); err != nil {
 		if errors.IsNotFound(err) {
-			log.Info("creating a new user", "name", userAcc.Name)
-			user = newUser(userAcc)
-			if err := controllerutil.SetControllerReference(userAcc, user, r.scheme); err != nil {
+			logger.Info("creating a new user", "name", userAcc.Name)
+			if err := r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, err.Error()); err != nil {
 				return nil, false, err
 			}
-			if err := r.client.Create(context.TODO(), user); err != nil {
-				return nil, false, errs.Wrapf(err, "failed to create user '%s'", userAcc.Name)
+			user = newUser(userAcc)
+			err = controllerutil.SetControllerReference(userAcc, user, r.scheme)
+			if err == nil {
+				err = r.client.Create(context.TODO(), user)
+			}
+			if err != nil {
+				return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, err, "failed to create user '%s'", userAcc.Name)
 			}
-			log.Info("user created successfully", "name", userAcc.Name)
+			logger.Info("user created successfully", "name", userAcc.Name)
 			return user, true, nil
 		}
-		return nil, false, errs.Wrapf(err, "failed to get user '%s'", userAcc.Name)
+		return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, err, "failed to get user '%s'", userAcc.Name)
 	}
-	log.Info("user already exists", "name", userAcc.Name)
+	logger.Info("user already exists", "name", userAcc.Name)
 	return user, false, nil
 }
 
-func (r *ReconcileUserAccount) ensureIdentity(userAcc *toolchainv1alpha1.UserAccount) (*userv1.Identity, bool, error) {
+func (r *ReconcileUserAccount) ensureIdentity(logger logr.Logger, userAcc *toolchainv1alpha1.UserAccount) (*userv1.Identity, bool, error) {
 	name := getIdentityName(userAcc)
 	identity := &userv1.Identity{}
 	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: name}, identity); err != nil {
 		if errors.IsNotFound(err) {
-			log.Info("creating a new identity", "name", name)
-			identity = newIdentity(userAcc)
-			if err := controllerutil.SetControllerReference(userAcc, identity, r.scheme); err != nil {
+			logger.Info("creating a new identity", "name", name)
+			if err := r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, err.Error()); err != nil {
```

Comment:
For happy scenario (first time reconcile for new user_account), this `updateStatus()` will set status=provisioning and error="user not found".  Assuming rest of the thing went well for create user and we returned.  Now, the status.error is still the same which is wrong.  We need to reset error to blank.  Below is the log where I am printing status on start of each reconcile().  You can see in second reconcile the status="provisioning" and error="user not found" but first reconcile run successfully with creating user.

```
{"level":"info","ts":1561619157.6916687,"logger":"controller_useraccount","msg":"Reconciling UserAccount","Request.Namespace":"toolchain-member-operator","Request.Name":"johnsmith"}
VN: u.s= { }
{"level":"info","ts":1561619157.6917093,"logger":"controller_useraccount","msg":"creating a new user","Request.Namespace":"toolchain-member-operator","Request.Name":"johnsmith","name":"johnsmith"}
{"level":"info","ts":1561619157.7047453,"logger":"controller_useraccount","msg":"user created successfully","Request.Namespace":"toolchain-member-operator","Request.Name":"johnsmith","name":"johnsmith"}
{"level":"info","ts":1561619157.8219233,"logger":"controller_useraccount","msg":"Reconciling UserAccount","Request.Namespace":"","Request.Name":"johnsmith"}
VN: u.s= {provisioning User.user.openshift.io "johnsmith" not found}
{"level":"info","ts":1561619157.8219843,"logger":"controller_useraccount","msg":"user already exists","Request.Namespace":"","Request.Name":"johnsmith","name":"johnsmith"}
{"level":"info","ts":1561619157.8220155,"logger":"controller_useraccount","msg":"creating a new identity","Request.Namespace":"","Request.Name":"johnsmith","name":"rhd:1a03ecac-7c0b-44fc-b66d-12dd7fb21c40"}
{"level":"info","ts":1561619157.8332338,"logger":"controller_useraccount","msg":"identity created successfully","Request.Namespace":"","Request.Name":"johnsmith","name":"rhd:1a03ecac-7c0b-44fc-b66d-12dd7fb21c40"}
{"level":"info","ts":1561619158.0225096,"logger":"controller_useraccount","msg":"Reconciling UserAccount","Request.Namespace":"","Request.Name":"johnsmith"}
```

Reply:
Good catch. I will address that and will add a test for that.

Reply:
Addressed in https://github.com/codeready-toolchain/member-operator/pull/25/commits/81f9d57c0d73f11d0430e1ae57b28891038d1a96